apm-agent-core/src/main/java/co/elastic/apm/bci/bytebuddy/CustomElementMatchers.java

```diff
@@ -0,0 +1,78 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.bci.bytebuddy;
+
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.nameStartsWith;
+import static net.bytebuddy.matcher.ElementMatchers.none;
+
+public class CustomElementMatchers {
+
+    public static ElementMatcher.Junction<TypeDescription> isInAnyPackage(Collection<String> includedPackages,
+                                                                          ElementMatcher.Junction<TypeDescription> defaultIfEmpty) {
+        if (includedPackages.isEmpty()) {
+            return defaultIfEmpty;
+        }
+        ElementMatcher.Junction<TypeDescription> matcher = none();
+        for (String applicationPackage : includedPackages) {
+            matcher = matcher.or(nameStartsWith(applicationPackage));
+        }
+        return matcher;
+    }
+
+    public static ElementMatcher.Junction<ClassLoader> classLoaderCanLoadClass(final String className) {
+        return new ElementMatcher.Junction.AbstractBase<ClassLoader>() {
+
+            private final boolean loadableByBootstrapClassLoader = canLoadClass(ClassLoader.getSystemClassLoader().getParent(), className);
```

Comment:
I believe `ClassLoader.getSystemClassLoader()` typically returns an instance (used to be `Luncher$AppClassLoader` before Java 9 modules and in my local JDK 10 returns an instance of `ClassLoaders$AppClassLoaders` in the open JDK), for which `getParent()` typically returns a non-null instance (used to be `Launcher$ExtClassLoader`; in my current JDK 10 I get `ClassLoaders$PlatformClassLoader`) which is not the bootstrap CL. If you invoke `getParent` on the latter you would get `null`, which indicates the bootstrap CL. So I believe the `canLoadClass` would yield `true` when it's false and NPE when it should be true.

Reply:
Two ideas:
* We could just require a `boolean` constructor argument whether or not the bootstrap CL should match
* call `getClass().getClassLoader()` on the class to check whether the returned CL is `null`

Reply:
oh, wait, `Class.forName` supports a `null` class loader